src/lib.rs

```diff
@@ -65,45 +72,94 @@ pub fn run_and_output(config: Config) {
 
 const MAX_FIX_ITERATIONS: usize = 10;
 
+#[allow(clippy::too_many_arguments)]
+fn run_per_file<'a>(
+    aggregated_queries: &'a AggregatedQueries<'a>,
+    path: &'a Path,
+    config: &'a Config,
+    mut on_found_violations: impl FnMut(Vec<ViolationWithContext>),
+    mut on_found_pending_fixes: impl FnMut(Vec<PendingFix>, &InstantiatedRule),
+    file_contents: impl Into<RopeOrSlice<'a>>,
+    tree: &'a Tree,
+    query: &'a Arc<Query>,
+    language: SupportedLanguage,
+    instantiated_rules: &'a [InstantiatedRule],
+) {
+    let file_contents = file_contents.into();
+    let mut instantiated_per_file_rules: HashMap<RuleName, Box<dyn RuleInstancePerFile<'a> + 'a>> =
+        Default::default();
+    get_matches(language.language(), file_contents, query, Some(tree)).for_each(|query_match| {
+        run_match(
+            query_match,
+            aggregated_queries,
+            language,
+            path,
+            file_contents,
+            config,
+            &mut instantiated_per_file_rules,
+            &mut on_found_violations,
+            |pending_fixes, instantiated_rule| {
+                on_found_pending_fixes(pending_fixes, instantiated_rule)
+            },
+        );
+    });
+
+    for (&instantiated_rule_index, &rule_listener_index) in
+        &aggregated_queries.instantiated_rule_root_exit_rule_listener_indices
```

Comment:
So far I've gotten a couple of the ESLint rules to work that included `*:exit` listeners by expressing that it alternate (slightly trickier in some cases) ways (I doubt that will be possible for all cases)

But one thing needed to support that was a final "cleanup" listener callback - in ESLint rules you can express that as `Program:exit` I believe but what I did was to wire up a "special" `ROOT_EXIT` (name?) listener that gets called after the querying has finished and passes you the root node

So here we're firing all of those that have been registered by the individual rules